### PR TITLE
Enforce the order of CMS creating ctrl calls

### PR DESCRIPTION
**Pull request by beldmit**
On encryption
1. Allocate the necessary EVP_PKEY_CTX
2. Set passed parameters, if any
3. Set information about the cipher used, if necessary
4. Call the ASN1 envelope ctrl

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
@slontis, @levitte - could you take a look, please?

**Comment by beldmit** [comment]
It's currently a semi-hack solution. To finalize it, we need to get rid of `CMS_KEY_PARAM` processing on encryption.

**Comment by beldmit** [comment]
Travis failures are relevant.

**Comment by beldmit** [comment]
Travis builds has succeeded all but 2 hanged ones.

File: crypto/evp/pmeth_lib.c [hunk 1]
```diff
@@ -747,7 +747,7 @@ static int legacy_ctrl_to_param(EVP_PKEY_CTX *ctx, int keytype, int optype,
                   EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
         return -2;
     }
-    return 0;
+    return -2;
```
Comment (slontis) [inline crypto/evp/pmeth_lib.c]: hmm.. It looks like the only way this function runs is when new provider code is running..
if ((EVP_PKEY_CTX_IS_DERIVE_OP(ctx) && ctx->op.kex.exchprovctx != NULL)
            || (EVP_PKEY_CTX_IS_SIGNATURE_OP(ctx)
                && ctx->op.sig.sigprovctx != NULL)
            || (EVP_PKEY_CTX_IS_ASYM_CIPHER_OP(ctx)
                && ctx->op.ciph.ciphprovctx != NULL))

So does it need to be a legacy ctrl?
Or could be just use EVP_PKEY_CTX_set_params() instead?

File: apps/cms.c [hunk 2]
```diff
@@ -877,23 +877,31 @@ int cms_main(int argc, char **argv)
         for (i = 0; i < sk_X509_num(encerts); i++) {
             CMS_RecipientInfo *ri;
             cms_key_param *kparam;
-            int tflags = flags;
+            int tflags = flags | CMS_KEY_PARAM; /* This flag enforces allocating the EVP_PKEY_CTX for the recipient here */
```
Comment (slontis) [inline apps/cms.c]: This top bit is not too different as it was doing this most of the time anyway.
Looks like the KeyAgree path should also work with this change as it doesnt use the flag.

File: apps/cms.c [hunk 3]
```diff
@@ -877,23 +877,31 @@ int cms_main(int argc, char **argv)
...
+
+            pctx = CMS_RecipientInfo_get0_pkey_ctx(ri);
             if (kparam != NULL) {
-                EVP_PKEY_CTX *pctx;
-                pctx = CMS_RecipientInfo_get0_pkey_ctx(ri);
                 if (!cms_set_pkey_param(pctx, kparam->param))
                     goto end;
             }
+
+            res = EVP_PKEY_CTX_ctrl(pctx, -1, -1,
```
Comment (slontis) [inline apps/cms.c]: See below comment - could we just use params here instead?

Comment (beldmit) [reply apps/cms.c]: Yes, EVP_PKEY_CTX_set_params seems relevant here too.

File: crypto/evp/pmeth_lib.c (hunk 1 above)
Comment (beldmit) [reply crypto/evp/pmeth_lib.c]: Yes, EVP_PKEY_CTX_set_params seems relevant here.

File: apps/cms.c (hunk 2 above)
Comment (beldmit) [reply apps/cms.c]: Here is the place where we process this bit:

https://github.com/openssl/openssl/blob/f64f26220442db3c6913188e6014e5bc5bc34653/crypto/cms/cms_env.c#L164-L171

If it is not set, we just call the default envelope ASN1 control. If it is, we first initialize the ctx, then set params and call the envelope ASN1 control that can refer to set parameters.

File: apps/cms.c (hunk 3 above)
Comment (beldmit) [reply apps/cms.c]: So what's the name of the parameter is relevant here?

**Comment by beldmit** [comment]
I've replaced EVP_PKEY_CTX_ctrl to EVP_PKEY_CTX_set_params but had to change the returning code of EVP_PKEY_CTX_set_params to -2 in case we don't have any appropriate settings. I'm not sure it's a good decision.

Any guidelines will be appropriated. @slontis, could you please look again?

File: apps/cms.c (hunk 3 above)
Comment (beldmit) [reply apps/cms.c]: Fixed

File: crypto/evp/pmeth_lib.c (hunk 1 above)
Comment (beldmit) [reply crypto/evp/pmeth_lib.c]: Rolled back

**Comment by beldmit** [comment]
It's also worth mentioning that EVP_PKEY_CTX_set_params is incompatible with ENGINE interfaces, and I'm not sure that all the engines will be converted to providers very fast...

File: crypto/evp/pmeth_lib.c [hunk 4]
```diff
@@ -588,7 +588,7 @@ int EVP_PKEY_CTX_set_params(EVP_PKEY_CTX *ctx, OSSL_PARAM *params)
             && ctx->op.ciph.cipher->set_ctx_params != NULL)
         return ctx->op.ciph.cipher->set_ctx_params(ctx->op.ciph.ciphprovctx,
                                                      params);
-    return 0;
```
Comment (slontis) [inline crypto/evp/pmeth_lib.c]: I dont think this is a good idea.. see above

File: apps/cms.c [hunk 5]
```diff
@@ -897,8 +900,10 @@ int cms_main(int argc, char **argv)
                     goto end;
             }
 
-            res = EVP_PKEY_CTX_ctrl(pctx, -1, -1,
-                        EVP_PKEY_CTRL_CIPHER, EVP_CIPHER_nid(cipher), NULL);
+            ctx_params[0] = OSSL_PARAM_construct_utf8_string("cipher",
+                cipher_name, strlen(cipher_name));
+            ctx_params[1] = OSSL_PARAM_construct_end();
+            res = EVP_PKEY_CTX_set_params(pctx, ctx_params);
             if (res <= 0 && res != -2)
```
Comment (slontis) [inline apps/cms.c]: You can query if its settable via..
EVP_PKEY_CTX_settable_params

File: apps/cms.c [hunk 6]
```diff
@@ -897,8 +900,10 @@ int cms_main(int argc, char **argv)
                     goto end;
             }
 
-            res = EVP_PKEY_CTX_ctrl(pctx, -1, -1,
-                        EVP_PKEY_CTRL_CIPHER, EVP_CIPHER_nid(cipher), NULL);
+            ctx_params[0] = OSSL_PARAM_construct_utf8_string("cipher",
```
Comment (slontis) [inline apps/cms.c]: This string would need to be defined properly as a #define and documented in a pod.

File: crypto/evp/pmeth_lib.c [hunk 7]
```diff
@@ -747,7 +747,7 @@ static int legacy_ctrl_to_param(EVP_PKEY_CTX *ctx, int keytype, int optype,
                   EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
         return -2;
     }
-    return -2;
```
Comment (slontis) [inline crypto/evp/pmeth_lib.c]: probably the same here..

File: apps/cms.c (hunk 6 above)
Comment (beldmit) [reply apps/cms.c]: Will do.

File: apps/cms.c (hunk 5 above)
Comment (beldmit) [reply apps/cms.c]: Fix applied.

File: apps/cms.c (hunk 6 above)
Comment (beldmit) [reply apps/cms.c]: Hopefully, it's enough or should it still be documented?

File: crypto/evp/pmeth_lib.c (hunk 7 above)
Comment (beldmit) [reply crypto/evp/pmeth_lib.c]: Removed

**Comment by beldmit** [comment]
I think we need smth like EVP_PKEY_CTX_param_is_settable(EVP_PKEY_CTX *, const char *).

**Comment by beldmit** [comment]
Also not sure whether I should deal with OPENSSL_strdup and OPENSSL_free on error or just remove the const modifier from the result of EVP_CIPHER_name explicitly.

File: apps/cms.c [hunk 8]
```diff
@@ -908,13 +912,15 @@ int cms_main(int argc, char **argv)
 
             settable = EVP_PKEY_CTX_settable_params(pctx);
```
Comment (slontis) [inline apps/cms.c]: An example is in evp_enc.c.. 
       /* Check the cipher actually understands this parameter */
        if (OSSL_PARAM_locate_const(EVP_CIPHER_settable_ctx_params(c->cipher),
                                    OSSL_CIPHER_PARAM_KEYLEN) == NULL)
            return 0;

Comment (slontis) [reply apps/cms.c]: There is also a OSSL_PARAM_locate

**Comment by slontis** [comment]
> Also not sure whether I should deal with OPENSSL_strdup and OPENSSL_free on error or just remove the const modifier from the result of EVP_CIPHER_name explicitly.

Depends on how it gets used.. It looks like it is not stored - it is just used to resolve a cipher and then thrown away - so the cast should be ok.. Maybe a comment saying this and a cast is ok.

File: apps/cms.c (hunk 8 above)
Comment (beldmit) [reply apps/cms.c]: Applied. Many thanks for the clue!

**Comment by beldmit** [comment]
CI is happy now.

**Comment by beldmit** [comment]
@slontis, could you please look at it now?

File: apps/cms.c [hunk 9]
```diff
@@ -877,23 +878,38 @@ int cms_main(int argc, char **argv)
...
+            /* As we do not get parameter here, the cast is safe */
+            ctx_params[0] = OSSL_PARAM_construct_utf8_string(
+                OSSL_ALG_PARAM_CIPHER, (char *)EVP_CIPHER_name(cipher), 
+                strlen(EVP_CIPHER_name(cipher)));
+            ctx_params[1] = OSSL_PARAM_construct_end();
+
+            if (OSSL_PARAM_locate_const(EVP_PKEY_CTX_settable_params(pctx),
+                OSSL_ALG_PARAM_CIPHER) != NULL) {
+                res = EVP_PKEY_CTX_set_params(pctx, ctx_params);
+                if (res <= 0 && res != -2)
```
Comment (slontis) [inline apps/cms.c]: The -2 check should no longer be needed..

**Comment by beldmit** [comment]
Travis is happy.

**Comment by beldmit** [comment]
@slontis, could you please re-review and approve if there are no problems with this patch?

**Comment by beldmit** [comment]
ping @openssl/otc

**Comment by beldmit** [comment]
See #10916 as full implementation of Russian GOST CMS.

**Comment by t8m** [comment]
Ping @beldmit  this is ready-to-merge

**Comment by beldmit** [comment]
This PR contradicts to the #10904 suggesting more comprehensive approach. So I'd prefer to leave unmerged for some time.

**Comment by levitte** [comment]
Ok, what's the state of this PR now?

**Comment by beldmit** [comment]
It is mostly superseded by #10904 so if you could look at it, it will be great.

**Comment by beldmit** [comment]
As #10904 is merged, this PR may be closed.

**Pull request closed without merging**